Eliminate Observable circular reference errors via lazier TypeRefinementContext building

DougGregor opened the pull request:
Eager expansion of type refinement contexts (TRCs) for variables within pattern binding declarations is causing cyclic references in some places involving macros. Make this expansion lazy, triggered by walking into these pattern binding declarations as part of (e.g.) availability queries. Also be careful to avoid "has property wrapper" queries as part of building the type refinement context, because they tend to cause circularities.

Fixes rdar://112079160, a case where we would see circular references when extending an `@Observable` class from another source file.

DougGregor committed:
[Type refinement context] Don't query property wrappers just for range info

Querying property wrappers involves semantic analysis that can cause
cyclic references while building the type refinement context, and it's
unnecessary: we need only know that these are custom attributes to
incorporate their source ranges. Switch to the simpler/cheaper query.

A small part of fixing the cyclic references in rdar://112079160.

DougGregor committed:
[Type refinement context] Lazily expand TRCs for pattern bindings

Eager expansion of type refinement contexts (TRCs) for variables
within pattern binding declarations is causing cyclic references in
some places involving macros. Make this expansion lazy, triggered by
walking into these pattern binding declarations as part of (e.g.)
availability queries.

Another step toward fixing the cyclic references in rdar://112079160.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor committed:
Make `VarDecl::isLayoutExposedToClients` check property wrappers more lazily

The check for "has property wrappers" as part of determining whether the
layout of a variable is exposed to clients can trigger reference
cycles. Push this check later, which eliminates these cycles for types
that aren't frozen/fixed-layout.

This is a hack, not a real fix, but it eliminates the cyclic
references observed in rdar://112079160.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Type refinement context] Avoid creating implicit contexts with bad ranges

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

tshortli commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -706,28 +717,79 @@ class TypeRefinementContextBuilder : private ASTWalker {
+    // the property wrapper attribute and use its source range to create a
+    // TRC for the initializer expression.
+    //
+    // FIXME: Since we don't have an expression here, we can't build out its
```
Is this comment no longer true, since we can now lazily expand the TRC for the init expr? I'm having a little trouble following whether this is handled now, but it sounds like it might be from a comment in the request implementation.

DougGregor replied:
I don't know that this has fully addressed the issue, because it seems like we'd need to add the TRCs for the property wrapper attributes (below). I'd need to dig into rdar://77841331 to see what it would take to fix that.

DougGregor replied:
Aside from the bugs I've added, I suspect this is related... this fails to type-check both before and after my PR:

```swift
@available(macOS 10.15, *)
func f() -> Int { 17 }

class X {
  @available(macOS 10.14, *)
  private static let value: Int = {
    if #available(macOS 10.15, *) {
      return f() // produces an error here saying that f() is not available
    }

    return 0
  }()
}
```

tshortli replied:
Ok, thanks for checking. I'll take a look at fixing that eventually!

DougGregor replied:
I changed the approach a little bit to in [this commit](https://github.com/apple/swift/pull/67642/commits/51b7b4d9658f793464f7041d912f54de3498c50a) deal with a failure in the source compatibility suite, which also fixes the example I gave above---an existing bug that's now fixed by my PR. It doesn't address rdar://77841331, though

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor committed:
Add test case involving circular references with `@Observable`

Add a test case for Observable types that are extended from other
source files. Prior to the recent changes to make
`TypeRefinementContext` more lazy, this would trigger circular
references through the `TypeRefinementContextBuilder`.

Finishes rdar://112079160.

DougGregor committed:
Drop unnecessary "parent context" state from TypeRefinementContextBuilder

This state is a holdover from when accessors we stored "alongside"
their variable declarations, rather than contained within them. That's
no longer the case, so we don't need to track this information any
more.

DougGregor committed:
Establish type refinement contexts for pattern binding decls directly

The type refinement context builder had a bunch of logic to try to
model type refinement contexts for the first variable declaration that
shows up within a pattern binding declaration. Instead, model this
more syntactically by creating a type refinement context for the
pattern binding declaration itself. This both addresses a regression
in the handling of `if #available` within a closure that's part of an
initializer, and fixes a bug in the same area where similar code has
explicit availability annotations.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented:
@swift-ci please test source compatibility release

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented:
@swift-ci please smoke test Windows

DougGregor commented:
@swift-ci please test source compatibility release

DougGregor commented:
@swift-ci please smoke test Windows

DougGregor commented:
@swift-ci please smoke test

tshortli commented on lib/AST/Availability.cpp:
```diff
@@ -749,3 +751,31 @@ ASTContext::getSwift5PlusAvailability(llvm::VersionTuple swiftVersion) {
 }
+
+const Decl *
+swift::abstractSyntaxDeclForAvailableAttribute(const Decl *ConcreteSyntaxDecl) {
```
Interesting, it feels like there's some overlap here with the `parentDeclForInferredAvailability` function that I introduced. I don't know that the two can merge, but maybe `parentDeclForInferredAvailability` should call this?

tshortli commented on test/Sema/availability_versions.swift:
```diff
@@ -1741,3 +1741,25 @@ func useHasUnavailableExtension(_ s: HasUnavailableExtension) {
+func f() -> Int { 17 }
+
+class StoredPropertiesWithAvailabilityInClosures {
+  private static let value: Int = {
```
This could maybe use a test case asserting that we diagnose a reference to `f` when there is no `if #available`.

DougGregor replied:
Oh, we actually don't diagnose a reference to `f` there because it's an initializer, and those aren't exposed to the client (so they're checked at the deployment target)

tshortli replied:
So then should f have availability later than the deployment target? Otherwise, we're not necessarily testing that TRCs get built out properly for the body.

DougGregor replied:
I have some test cases locally to check this, and will push them to a separate PR shortly.

tshortli commented on test/stdlib/Observation/ObservableAvailabilityCycle.swift:
```diff
@@ -0,0 +1,19 @@
+
+// RUN: %target-swift-frontend -typecheck -parse-as-library -enable-experimental-feature InitAccessors -external-plugin-path %swift-host-lib-dir/plugins#%swift-plugin-server -primary-file %s %S/Inputs/ObservableClass.swift
+
+// RUN: %target-swift-frontend -typecheck -parse-as-library -enable-experimental-feature InitAccessors -external-plugin-path %swift-host-lib-dir/plugins#%swift-plugin-server %s -primary-file %S/Inputs/ObservableClass.swift
```
Should this be building `%s` instead of `ObservableClass.swift` again?

tshortli replied:
Oh I see the `%s` now

compnerd commented:
This seems to be causing some trouble on CI for macOS:

> /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/test/stdlib/Observation/Inputs/ObservableClass.swift:5:2: error: 'Observable()' is only available in macOS 14.0 or newer
@Observable final public class ObservableClass {
 ^

slavapestov commented:
Instead of reverting why not just disable the new test?
